sw/device/lib/dif/dif_pwm.h

```diff
@@ -0,0 +1,322 @@
+ *  |                                                         |
+ *  v                                                         v
+ * |-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-| ... |-|-|-|-|
+ *
+ * |-|
+ * min beat size = 1 phase counter ticks --> 2^16 "beats" / "pulse cycle"
+ *
+ * |---------------------------------------------- ... | ... --|
+ *    max beat size = 2^15 phase counter ticks --> 2 "beats" / "pulse cycle"
+ */
+
+#include <stdint.h>
+
+#include "sw/device/lib/dif/autogen/dif_pwm_autogen.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Helper X macro for defining enums and case statements related to PWM
+ * channels. If an additional channel is ever added to the hardware, this list
+ * can be updated.
+ */
+#define LIST_OF_CHANNELS(X) \
+  X(0)                      \
+  X(1)                      \
+  X(2)                      \
+  X(3)                      \
+  X(4)                      \
+  X(5)
+
+/**
+ * Helper macro for defining a `dif_pwm_channel_t` enumeration constant.
+ * @channel_ PWM channel of the enumeration constant.
+ */
+#define PWM_CHANNEL_ENUM_INIT_(channel_) \
+  kDifPwmChannel##channel_ = 1 << channel_,
+
+/**
+ * A PWM channel.
+ */
+typedef enum dif_pwm_channel {
+  LIST_OF_CHANNELS(PWM_CHANNEL_ENUM_INIT_)
+} dif_pwm_channel_t;
+
+/**
+ * Runtime configuration for PWM.
+ *
+ * This struct describes runtime configuration for one-time configuration of the
+ * PWM "pulse cycle" and "beat" durations that impact all PWM channels.
+ */
+typedef struct dif_pwm_config {
+  /**
+   * The core clock frequency divisor that determines the period of a single
+   * "beat" within a PWM "pulse cycle".
+   *
+   * Valid range: [0, 2^26)
+   *
+   * A value of zero, configures the "beat" period to the core clock period.
+   */
+  uint32_t clock_divisor;
+  /**
+   * The total number of "beats" in a "pulse cycle", including both "on" and
+   * "off" beats in a "pulse cycle".
+   *
+   * Valid range: [2, 2^16]
+   *
+   * A "beat" represents a unit of time of the PWM output signal. Higher values
+   * provide higher duty cycle resolutions, at the expense of longer "pulse
+   * cycles", while lower values provide shorter "pulse cycles", at the expense
+   * of lower duty cycle resolutions (since duty cycles are configured in
+   * "beats" / "pulse cycle").
+   */
+  uint32_t beats_per_pulse_cycle;
+} dif_pwm_config_t;
+
+/**
+ * A PWM channel mode.
+ */
+typedef enum dif_pwm_mode {
+  /**
+   * The PWM duty cycle is set by the firmware and remains constant.
+   */
+  kDifPwmModeFirmware = 0,
+  /**
+   * The PWM duty cycle linearly sweeps between both primary and secondary
+   * firmware-configured values, based on a firmware-configured step size.
+   */
+  kDifPwmModeHeartbeat = 1,
+  /**
+   * The PWM duty cycle alternates between both primary and secondary
+   * firmware-configured values, based on two firmware-configured durations.
+   */
+  kDifPwmModeBlink = 1,
```

Comment:
This is a repeat of heartbeat, which I assume is a typo.

Reply:
Fixed.